---
title: "James jmcne R Homework"
output: html_notebook
---

```{r}
getwd()
setwd("C:/Users/jmcne/Dropbox/Spring 2017/BCB 546X/R")
fang<-read.table("C:/Users/jmcne/Dropbox/Spring 2017/BCB 546X/R/fang_et_al_genotypes.txt",header=T)
str(fang)
fang[1:10, 1:10]
fang[]<-lapply(fang, as.character) #making every characters, needed for next step
fang<-rbind(colnames(fang),fang) #making first row column names, will save a step later when I transpose the fang df

snps<-read.delim("C:/Users/jmcne/Dropbox/Spring 2017/BCB 546X/Unix/hw1/snp_position.txt")
str(snps)
snps[1:10, 1:10]
snps$Chromosome<-as.numeric(as.character(snps$Chromosome)) #ignore warnings
snps$Position<-as.numeric(as.character(snps$Position)) #ignore warnings

#need to merge them together by SNP_ID, first need to get 'fang' df formatted so SNP_ID is in column 1
fang_t<-as.data.frame(t(fang),row.names=T)
fang_t[1:5, 1:5]
colnames(fang_t)<-as.character(unlist(fang_t[1,]))
all_groups<-fang_t[3,]
fang_t<-fang_t[-c(1:3),] #for part 1
#write.csv(all_groups, "list of all groups.csv",row.names=F) #for part 2
colnames(fang_t)[1]<-paste("SNP_ID")

#formatting SNPs
snps_2<-snps[,c(1,3,4)]
snps_2[1:10,]
str(snps_2)

#merging
combined<-merge(snps_2, fang_t, by="SNP_ID")
combined[,-c(1:3)]<-lapply(combined[,-c(1:3)], as.character) #making SNP calls characters, required for making ?/? into -/- later
combined[1:10, 1:10]
write.csv(combined, "combined data all.csv", row.names=F)


#subsetting, just maize groups of interest
maize<-cbind(combined[,c(1:3)],combined[,which(colnames(combined) %in% (as.character(fang[which(fang$Group %in% list("ZMMIL","ZMMLR","ZMMMR")),1])))])
maize<-maize[order(maize[,2], maize[,3]),]

#subsetting teosinte groups of interest
teosinte<-cbind(combined[,c(1:3)],combined[,which(colnames(combined) %in% (as.character(fang[which(fang$Group %in% list("ZMPBA","ZMPIL","ZMPJA")),1])))])
teosinte<-teosinte[order(teosinte[,2], teosinte[,3]),]
teosinte[1:10, 1:10]

#looping through maize dataframe by chromosome and saving output
for(i in 1:10){
  temp_df<-maize[maize[,2]==i,]
  write.csv(temp_df, paste("maize_chr",i,".csv",sep=""),row.names=F)
  temp_df[temp_df=="?/?"]<-"-/-"
  write.csv(temp_df, paste("maize_chr",i,"_altered_missing_values.csv",sep=""),row.names=F)
}

#looking through teosinte dataframe and saving output
for(i in 1:10){
  temp_df<-teosinte[teosinte[,2]==i,]
  write.csv(temp_df, paste("teosinte_chr",i,".csv",sep=""),row.names=F)
  temp_df[temp_df=="?/?"]<-"-/-"
  write.csv(temp_df, paste("teosinte_chr",i,"_altered_missing_values.csv",sep=""),row.names=F)
}
```



##Part II - graphs
```{r}
library("ggplot2")
#loading data from part I and formatting
combined<-read.csv("combined data all.csv")
combined<-combined[order(combined[,2], combined[,3]),]
combined[1:10, 1:10]
row.names(combined)<-seq(1,nrow(combined),1)
all_groups<-read.csv("list of all groups.csv",stringsAsFactors = F)
all_groups<-all_groups[-1]#formatting for future rbind
all_groups<-cbind(0,0,0,all_groups)
all_groups[1:10]
colnames(all_groups)<-colnames(combined)
combined<-rbind(all_groups, combined) #making group info a row, for a later loop
combined[1:10, 1:10]
row.names(combined)<-seq(1,nrow(combined),1)

groups<-as.character(unique(fang$Group))

#looping through and getting the number SNPs per chromosome
snps_per_chrom<-data.frame()
for(i in 1:10){
  temp<-nrow(combined[combined$Chromosome==i,])
  snps_per_chrom<-rbind(snps_per_chrom,cbind(i,temp))
  }
colnames(snps_per_chrom)<-c("Chromosome", "num_SNPs")

#looping through and getting number of SNPs per Group
by_group<-data.frame()
for(j in 1:length(groups)){
  temp<-ncol(combined[,combined[1,]==groups[j]])
  temp<-cbind(groups[j],temp)
  by_group<-rbind(by_group,temp)
}
colnames(by_group)<-c("group", "Num_Genotypes")
by_group[,2]<-as.numeric(as.character(by_group[,2]))

#plotting number of SNPs per chromosome
ggplot(snps_per_chrom, aes(x=snps_per_chrom$Chromosome,y=snps_per_chrom$num_SNPs))+geom_bar(stat="identity")+
  labs(list(x="Chromosome",y="# of SNPs", title="SNPs per Chromosome")) + 
  scale_x_discrete(limits=seq(1,10,1))

#plotting number of Genotypes per Group
ggplot(by_group, aes(x=by_group$group, y=by_group$Num_Genotypes)) + geom_bar(stat="identity") +
  labs(list(x="Group", y="# of Genotypes", title="Number of Genotypes per Group")) +
  theme(axis.text.x=element_text(angle=90))


#missing data and amount of heterzygosity
fang<-read.table("C:/Users/jmcne/Dropbox/Spring 2017/BCB 546X/R/fang_et_al_genotypes.txt",header=T)
fang[1:10, 1:10]
fang[]<-lapply(fang, as.character) #making data into characters, needed for next steps
fang<-fang[,-2]

fang[,-c(1:2)][fang[,-c(1:2)]=="?/?"]<-NA

fang<-fang[order(fang[,1], fang[,2]),]

fang2<-fang
#making homozygotes equal to 1, heterozygotes equal to 0
start_time1<-proc.time()
for(i in 1:nrow(fang2)){
  for(j in 3:ncol(fang2)){
    if(is.na(fang2[i,j])==FALSE){
      if(strsplit(fang2[i,j], "\\/")[[1]][1]==strsplit(fang2[i,j], "\\/")[[1]][2]){
        fang2[i,j]<-1
      }else if(strsplit(fang2[i,j], "\\/")[[1]][1]!=strsplit(fang2[i,j], "\\/")[[1]][2])
        fang2[i,j]<-0
    }
  }
}
time1<-proc.time()-start_time1
time1 #466.03, slow, but wrote it another way that took c. 750.

fang2[1:10, 1:8]

str(fang2)
fang2[,-c(1:2)]<-lapply(fang2[,-c(1:2)],as.numeric)

#function to get the percentage of homozygotes, heterozygotes and missing
num_col<-ncol(fang2[,-c(1:2)])
call_percents<- function(x){
  SNP<-unlist(x)
  percent_homozygotes<-sum(SNP, na.rm=T) / num_col
  percent_missing<- sum(is.na(SNP))/num_col
  percent_heterzygotes<-1-percent_homozygotes-percent_missing
  out<-cbind(percent_homozygotes, percent_heterzygotes,percent_missing)
}

fang_percentages<-as.data.frame(apply(fang2[,-c(1:2)],1,call_percents)) #applying function
dim(fang_percentages)
fang_percentages<-t(fang_percentages)
head(fang_percentages)
colnames(fang_percentages)<-c("percent_homozygotes", "percent_heterzygotes", "percent_missing")
fang_percentages<-cbind(fang2[,c(1:2)], fang_percentages)

library("reshape2")
fang_percentages2<-melt(fang_percentages, id=c("Sample_ID", "Group"))
head(fang_percentages2)
#plotting the breakdown of SNP calls by Genotype (species)
ggplot(fang_percentages2, aes(x=fang_percentages2$Sample_ID, y=fang_percentages2$value, fill=fang_percentages2$variable)) + 
  geom_bar(stat="identity") +
 labs(list(x="Genotype",y="Percent of SNP calls", title="Percentage of SNPs calls, grouped by Genotype")) + 
  theme(axis.text.x=element_blank(), plot.title=element_text(hjust=.5)) + guides(fill=guide_legend((title="Legend"))) 


#Getting the percentage of heterozygotes, homozygotes and missing per group
groups<-unique(fang_percentages$Group)
group_summary<-data.frame()
for(i in 1:length(groups)){
  temp<-fang_percentages[fang_percentages$Group==groups[i],]
  group_summary<-rbind(group_summary, cbind(groups[i], mean(temp[,3]), mean(temp[,4]), mean(temp[,5])))
}
colnames(group_summary)<-c("group", "homozygotes", "heterozygotes", "missing")
group_summary[,2]<-as.numeric(as.character(group_summary[,2]))
group_summary[,3]<-as.numeric(as.character(group_summary[,3]))
group_summary[,4]<-as.numeric(as.character(group_summary[,4]))
group_summary<-melt(group_summary, id="group")
head(group_summary);str(group_summary)

#plotting the breakdown of SNP calls by Group
ggplot(group_summary, aes(x=group_summary$group, y=group_summary$value, fill=group_summary$variable)) + geom_bar(stat ="identity", position="dodge") +
  labs(list(x="Group",y="Percent of SNP calls", title="Percentage of SNPs calls by Group")) + 
  theme(plot.title=element_text(hjust=.5), axis.text.x=element_text(angle=45)) + guides(fill=guide_legend((title="Legend")))

##########################################################################
###Visualize one other feature of the dataset. The choice is up to you!###
##########################################################################

#getting the number of non-NA SNPs for each group within each chromosome
group_within_chr<-data.frame()
for(i in 1:10){
  tempdf<-combined[combined$Chromosome==i,]
  tempdf<-rbind(combined[1,],tempdf)
  temp_none_NA_snps<-((nrow(tempdf)-1) * (ncol(tempdf)-3)) - sum(sapply(tempdf[-1,-c(1:3)], function(x) sum(is.na(x)))) #total number non-NA SNPs in ith chromosome
  for(j in 1:length(groups)){
    temp_group_df<-tempdf[,tempdf[1,]==groups[j]]
    group_within_chr<-rbind(group_within_chr, cbind(i,groups[j],((((nrow(temp_group_df)-1) * (ncol(temp_group_df))) - sum(sapply(temp_group_df, function(x) sum(is.na(x))))) / temp_none_NA_snps)))
  }
}
colnames(group_within_chr)<-c("chr", "group", "percent_SNPs")
head(group_within_chr)
str(group_within_chr)
group_within_chr[,3]<-as.numeric(as.character(group_within_chr[,3]))
group_within_chr[,1]<-as.numeric(as.character(group_within_chr[,1]))
group_within_chr[,2]<-as.character(levels(group_within_chr[,2]))[group_within_chr[,2]]

#can not really tell a different when looking at the graphs in R notebook, when zoomed in there is a very, very slight change in frequency of non-missing SNPs across chromosomes for one or two groups.  
#Number of non-NA SNPs per group, stacked barchart
ggplot(group_within_chr, aes(x=group_within_chr$chr, y=group_within_chr$percent_SNPs, fill=group_within_chr$group)) + geom_bar(stat="identity") +
  labs(list(x="Chromosome",y="Non-Missing SNPs per Group", title="Percent Non-Missing SNPs by Group")) + scale_x_discrete(limits=seq(1,10,1)) + guides(fill=guide_legend((title="Legend")))

#Number of non-NA SNPs per group, line graph
ggplot(group_within_chr, aes(x=group_within_chr$chr, y=group_within_chr$percent_SNPs)) + geom_line(aes(color=group_within_chr$group))  +
  labs(list(x="Chromosome",y="Non-Missing SNPs per Group", title="Percent Non-Missing SNPs by Group")) + scale_x_discrete(limits=seq(1,10,1)) + guides(fill=guide_legend((title="Legend")))


```
